File: utils/images/formats.py
"""Image identification based on actual file contents."""

from io import BytesIO
from PIL import Image, ImageOps

MIME_TYPES = {
    "jpg": "image/jpeg",
    "png": "image/png",
    "gif": "image/gif",
    "webp": "image/webp",
    "bmp": "image/bmp",
}


def extension_for_mime(content_type: str):
    mime = content_type.split(";", 1)[0].strip().lower()
    if mime == "image/jpg":
        mime = "image/jpeg"
    return next(
        (extension for extension, value in MIME_TYPES.items() if value == mime), None
    )


def detect_format(data: bytes) -> str:
    if data.startswith(b"\xff\xd8\xff"):
        return "jpg"
    if data.startswith(b"\x89PNG\r\n\x1a\n"):
        return "png"
    if data.startswith((b"GIF87a", b"GIF89a")):
        return "gif"
    if data[:4] == b"RIFF" and data[8:12] == b"WEBP":
        return "webp"
    if data.startswith(b"BM"):
        return "bmp"
    raise ValueError("Unknown image format")


def validate_image(data: bytes) -> str:
    extension = detect_format(data)
    with Image.open(BytesIO(data)) as image:
        image.verify()
    return extension


def to_rgb(image):
    if image.mode in ("RGBA", "LA") or "transparency" in image.info:
        rgba = image.convert("RGBA")
        return Image.alpha_composite(
            Image.new("RGBA", rgba.size, "white"), rgba
        ).convert("RGB")
    return image.convert("RGB")


def image_to_bytes(image, fmt="PNG", **options):
    buffer = BytesIO()
    image.save(buffer, format=fmt, **options)
    return buffer.getvalue()

File: utils/images/formats.go
package images

import (
	"bytes"
	"errors"
	"image"
	"image/color"
	"image/draw"
	"image/gif"
	"image/jpeg"
	"image/png"
	"strings"

	"golang.org/x/image/bmp"
	_ "golang.org/x/image/webp"
)

func ExtensionForMIME(mime string) string {
	switch strings.TrimSpace(strings.Split(mime, ";")[0]) {
	case "image/jpeg", "image/jpg":
		return "jpg"
	case "image/png":
		return "png"
	case "image/gif":
		return "gif"
	case "image/webp":
		return "webp"
	case "image/bmp":
		return "bmp"
	}
	return ""
}
func Decode(data []byte) (image.Image, string, error) {
	img, format, err := image.Decode(bytes.NewReader(data))
	if format == "jpeg" {
		format = "jpg"
	}
	return img, format, err
}
func Validate(data []byte) (string, error) { _, format, err := Decode(data); return format, err }
func RGB(img image.Image) *image.RGBA {
	out := image.NewRGBA(image.Rect(0, 0, img.Bounds().Dx(), img.Bounds().Dy()))
	draw.Draw(out, out.Bounds(), image.NewUniform(color.White), image.Point{}, draw.Src)
	draw.Draw(out, out.Bounds(), img, img.Bounds().Min, draw.Over)
	return out
}
func Encode(img image.Image, format string) ([]byte, error) {
	var b bytes.Buffer
	var err error
	switch strings.ToLower(format) {
	case "png", "":
		err = png.Encode(&b, img)
	case "jpg", "jpeg":
		err = jpeg.Encode(&b, RGB(img), &jpeg.Options{Quality: 95})
	case "gif":
		err = gif.Encode(&b, img, nil)
	case "bmp":
		err = bmp.Encode(&b, img)
	default:
		err = errors.New("unsupported output image format")
	}
	return b.Bytes(), err
}
